src/mame/drivers/firebeat.cpp

```diff
@@ -154,14 +155,217 @@
 #include "osdcomm.h"
 
+#include "wdlfft/fft.h"
+
+#include <cmath>
+
 #include "firebeat.lh"
 
 
+/*****************************************************************************/
+DECLARE_DEVICE_TYPE(KONAMI_FIREBEAT_EXTEND_SPECTRUM_ANALYZER, firebeat_extend_spectrum_analyzer_device)
+
+class firebeat_extend_spectrum_analyzer_device : public device_t, public device_mixer_interface
+{
+public:
+	firebeat_extend_spectrum_analyzer_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	uint8_t read(offs_t offset);
+
+protected:
+	// device-level overrides
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	// device_sound_interface-level overrides
+	void sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs) override;
+
+private:
+	enum {
+		TOTAL_BUFFERS = 2,
+		TOTAL_CHANNELS = 2,
+		TOTAL_BARS = 6,
+
+		FFT_LENGTH = 512
+	};
+
+	void update_fft();
+	void apply_fft(uint32_t buf_index);
+
+	float m_audio_buf[TOTAL_BUFFERS][TOTAL_CHANNELS][FFT_LENGTH];
+	float m_fft_buf[TOTAL_CHANNELS][FFT_LENGTH];
+	int m_audio_fill_index;
+	int m_audio_count[TOTAL_CHANNELS];
+
+	int m_bars[TOTAL_CHANNELS][TOTAL_BARS];
+};
+
+firebeat_extend_spectrum_analyzer_device::firebeat_extend_spectrum_analyzer_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, KONAMI_FIREBEAT_EXTEND_SPECTRUM_ANALYZER, tag, owner, clock),
+	device_mixer_interface(mconfig, *this, 2)
+{
+}
+
+void firebeat_extend_spectrum_analyzer_device::device_start()
+{
+	WDL_fft_init();
+}
+
+void firebeat_extend_spectrum_analyzer_device::device_reset()
+{
+	for (int ch = 0; ch < TOTAL_CHANNELS; ch++)
+	{
+		for (int i = 0; i < TOTAL_BUFFERS; i++) {
+			std::fill(std::begin(m_audio_buf[i][ch]), std::end(m_audio_buf[i][ch]), 0);
+		}
+
+		std::fill(std::begin(m_fft_buf[ch]), std::end(m_fft_buf[ch]), 0);
+
+		m_audio_count[ch] = 0;
+
+		for (int i = 0; i < TOTAL_BARS; i++)
+		{
+			m_bars[ch][i] = 0;
+		}
+	}
+
+	m_audio_fill_index = 0;
+}
+
+void firebeat_extend_spectrum_analyzer_device::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
+{
+	device_mixer_interface::sound_stream_update(stream, inputs, outputs);
+
+	for (int pos = 0; pos < outputs[0].samples(); pos++)
+	{
+		for (int ch = 0; ch < outputs.size(); ch++)
+		{
+			const float sample = outputs[ch].get(pos);
+			m_audio_buf[m_audio_fill_index][ch][m_audio_count[m_audio_fill_index]] = sample;
+		}
+
+		update_fft();
+	}
+
+	// Bandpass filter + find peak
+	// The results aren't accurate to the real circuit used on the PCB but the feature is very minor
+	// and does not affect gameplay in any way.
+
+	// Band values taken directly from NJU7507 data sheet.
+	double notches[] = { 95, 240, 600, 1500, 3400, 8200, 18000 };
+	int last_notch = std::end(notches) - std::begin(notches);
```

Comment:
Same with this – seems it’s constant, and could use `std::size` rather than the iterator subtraction?

Reply:
A constant is perfectly fine here. I've changed both to `constexpr` now.